app/main/presenters/search_presenters.py

```diff
@@ -72,3 +89,70 @@ def set_filter_states(filter_groups, request):
                 filter['checked'] = (
                     filter['value'] in param_values
                 )
+
+
+def show_lots_and_categories_selection(lots, category_filter_group, request):
+    """
+    Equivalent of set_filter_states but for where we are creating a tree of links i.e. the
+    lots/categories widget. Adds links (where necessary) and shows the currently-selected
+    lot and/or categories.
+    """
+    current_lot_slug = get_lot_from_request(request)
+
+    for lot in lots:
+        lot_selected = (lot['slug'] == current_lot_slug)
+        if lot_selected:
+            lot['selected'] = True
+            categories = category_filter_group['filters'] if category_filter_group else []
+            category_was_selected = _show_category_selection(lot, categories, request)
+            lot['categories'] = categories
+
+        if not lot_selected or category_was_selected:
+            # we need a link back to the lot _without_ a category selected
+            lot['link'] = url_for('.search_services', q=get_keywords_from_request(request), lot=lot['slug'])
+
+
+def _show_category_selection(lot, category_filters, request):
+    """
+    Recursive setting of 'selected' state and adding of links to categories
+    as part of building the lots/categories tree.
+    :param category_filters: set of category filters as previously produced by filters_for_question
+    :param request: request object from which to extract active filters
+    :return: True iff a category from the list was selected
```

Comment:
I was taught to do this, you are not the first to be confused, but I would find it difficult to stop now :-(

"iff" === "if and only if", i.e. "True if a category from the list was selected, but otherwise False", in other words, extremely pedantic logical ass-covering.